openhands/controller/agent_controller.py

```diff
@@ -133,21 +131,16 @@ async def update_state_after_step(self):
         # update metrics especially for cost. Use deepcopy to avoid it being modified by agent.reset()
         self.state.local_metrics = copy.deepcopy(self.agent.llm.metrics)
 
-    async def report_error(self, message: str, exception: Exception | None = None):
-        """Reports an error to the user and sends the exception to the LLM next step, in the hope it can self-correct.
-
-        This method should be called for a particular type of errors, which have:
-        - a user-friendly message, which will be shown in the chat box. This should not be a raw exception message.
-        - an ErrorObservation that can be sent to the LLM by the user role, with the exception message, so it can self-correct next time.
-        """
-        self.state.last_error = message
-        if exception:
-            self.state.last_error += f': {exception}'
-        detail = str(exception) if exception is not None else ''
-        if exception is not None and isinstance(exception, litellm.AuthenticationError):
-            detail = 'Please check your credentials. Is your API key correct?'
-        self.event_stream.add_event(
-            ErrorObservation(f'{message}:{detail}'), EventSource.USER
+    async def _react_to_error(
+        self,
+        message: str,
+        new_state: AgentState | None = None,
+    ):
+        if new_state is not None:
+            # it's important to set the state before adding the error event, so that metrics sync properly
+            await self.set_agent_state_to(new_state)
+        await self.event_stream.async_add_event(
+            ErrorObservation(message), EventSource.AGENT
```

Comment:
Are you sure about changing USER to AGENT? We were just discussing [about this](https://github.com/All-Hands-AI/OpenHands/pull/4387#discussion_r1800341181).

Reply:
This might not matter much with tool use, because that one is `role=tool`! Or I hope it stays role tool, anyway. (I'm not sure yet if/how all tool use APIs support the role)

Reply:
_Edited to add:_ oh! For the UI we need it AGENT, right? Well... so this is the mix UI/LLM in ErrorObservations hitting. 🤔

We could adjust it in the agent for obs... but if that's what you want to go with, please lets add some comment. We _will_ forget, because some of us reflexively assume that the LLM's way is _the way things should be_. 😂

Maybe that idea linked above, to have in our code an "environment" (or "tool") role, is not too bad.

Reply:
Thanks for solving this one!